midterm-2023-11-06/code/5100-nenad-saldziev/script.ts

```diff
@@ -0,0 +1,188 @@
+interface albums{ 
+        name: string; // the name of the album
+        year: number; // the year the album was released
+} // an array of album objects
+
+interface metalBand{  // Author == metalBand
+    id: number;
+    name: string;
+    genre: string;
+    formed: number; // the year the band was formed
+    location : string; // the city and country where the band was formed
+    members: [ string ]; // an array of strings, each string is the name of a band member
+    album: albums[];
+}
+// Array of band[objects]
+
+type BandSorter = (first: metalBand, second: metalBand) => number;
+
+document.addEventListener("DOMContentLoaded", backsitcode)
+
+let bands: metalBand[] = []; // all authors == bands
+
+async function backsitcode() {
+    const data = await waitData();
+    bands = data;
+
+    fillter(bands); // fillNationalities == fillter
+
+    displayBands(bands); // displayAuthors == displayBands
+
+    const nameSort = document.getElementById("sort-name")!;
+    nameSort.addEventListener("click", sortByName); // sortByName == sortByName
+
+    const idSort = document.getElementById("sort-id")!;
+    idSort.addEventListener("click", sortById); // sortById == sortById
+
+    const applyFilterButton = document.getElementById("apply-filter")!;
+    applyFilterButton.addEventListener("click", applyFilter); // applyFilter == applyFilter
+
+    const modal = document.getElementById("biblio-details")!;
+    modal.addEventListener("click", () => {
+        modal.classList.add("hidden");
+    });
+}
+
+const nameSorter: BandSorter = (first, second) => first.name.localeCompare(second.name);
+const idSorter: BandSorter = (first, second) => first.id - second.id;
+
+const sortByName = () => {
+    const sortedBands = bands.toSorted(nameSorter);
+    displayBands(sortedBands);
+}
+
+const sortById = () => {
+    const sortedBands = bands.toSorted(idSorter);
+    displayBands(sortedBands);
+}
+
+const fillter = (bands: metalBand[]) => {
+    const filter = document.getElementById("country-filter")!;
+
+    const countrys = new Set<string>();
+    for (const band of bands) {
+        countrys.add(band.location);
+    }
+
+    for (const country of countrys) {
+        const option = document.createElement("option");
+        option.value = country;
+        option.innerHTML = country;
+        filter.appendChild(option);
+    }
+}
+
+const applyFilter = () => {
+    const countryElement = document.getElementById("country-filter") as HTMLSelectElement;
+    const country = countryElement.value;
+
+    const aliveElement = document.getElementById("alive-filter") as HTMLSelectElement;
+    const alive = aliveElement.value;
+
+    let filteredBands = bands;
+    if (country !== "all") {
```

Comment:
Given the locations logic in the population, this will work for specific locations - not countries. The "alive" logic is flawed, it's impossible for a band not to have a formed date :) +10